BoxContentSDK/BoxContentSDK/BOXNSURLSessionManager.m

```diff
@@ -45,49 +57,145 @@ - (NSURLSession *)defaultSession
+        [self.sessionTaskIdToTaskDelegate setObject:taskDelegate forKey:@(sessionTaskId)];
+    }
+}
+
+- (void)dessociateSessionTaskId:(NSUInteger)sessionTaskId
+{
+    @synchronized (self.sessionTaskIdToTaskDelegate) {
+        [self.sessionTaskIdToTaskDelegate removeObjectForKey:@(sessionTaskId)];
     }
-    return _sessionDelegate;
 }
 
 - (NSURLSessionDataTask *)createDataTask:(NSURLRequest *)request completionHandler:(void (^)(NSData * data, NSURLResponse * response, NSError * error))completionHandler
 {
     return [self.defaultSession dataTaskWithRequest:request completionHandler:completionHandler];
 }
 
-- (NSURLSessionDownloadTask *)createDownloadTaskWithRequest:(NSURLRequest *)request
+- (NSURLSessionDataTask *)createDataTaskForDownload:(NSURLRequest *)request taskDelegate:(id <BOXNSURLSessionTaskDelegate, BOXNSURLSessionDownloadTaskDelegate>)taskDelegate
+{
+    NSURLSessionTask *task = [self.advancedSession dataTaskWithRequest:request];
+    [self associateSessionTaskId:task.taskIdentifier withTaskDelegate:taskDelegate];
+    return task;
+}
+
+- (NSURLSessionDownloadTask *)createDownloadTaskWithRequest:(NSURLRequest *)request taskDelegate:(id <BOXNSURLSessionTaskDelegate, BOXNSURLSessionDownloadTaskDelegate>)taskDelegate
 {
-    return [self.backgroundSession downloadTaskWithRequest:request];
+    NSURLSessionTask *task = [self.advancedSession downloadTaskWithRequest:request];
+    [self associateSessionTaskId:task.taskIdentifier withTaskDelegate:taskDelegate];
+    return task;
 }
 
 - (NSURLSessionDownloadTask *)createDownloadTaskWithResumeData:(NSData *)resumeData
 {
-    return [self.backgroundSession downloadTaskWithResumeData:resumeData];
+    return [self.advancedSession downloadTaskWithResumeData:resumeData];
 }
 
 - (NSURLSessionUploadTask *)createUploadTask:(NSURLRequest *)request fromFile:(NSURL *)fileURL
 {
-    return [self.backgroundSession uploadTaskWithRequest:request fromFile:fileURL];
+    return [self.advancedSession uploadTaskWithRequest:request fromFile:fileURL];
+}
+
+#pragma mark - implementations for NSURLSession-related delegates for advanced session
+
+#pragma mark - NSURLSessionDownloadDelegate
+
+/* Sent when a download task that has completed a download.  The delegate should
+ * copy or move the file at the given location to a new location as it will be
+ * removed when the delegate message returns. URLSession:task:didCompleteWithError: will
+ * still be called.
+ */
+- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask
+didFinishDownloadingToURL:(NSURL *)location
+{
+    id<BOXNSURLSessionTaskDelegate> taskDelegate = [self.sessionTaskIdToTaskDelegate objectForKey:@(downloadTask.taskIdentifier)];
+    if ([taskDelegate conformsToProtocol:@protocol(BOXNSURLSessionDownloadTaskDelegate)]) {
+        id<BOXNSURLSessionDownloadTaskDelegate> downloadTaskDelegate = taskDelegate;
+
+        if ([downloadTaskDelegate respondsToSelector:@selector(destinationPath)]) {
```

Comment:
Should we also do a `nil` check on the destination path? This probably shouldn't happen, but you know, it's not guaranteed by the type. :)